File: frontend/src/types/qualitycontrol10.ts
/** Domain types for qualitycontrol / QualityControlRecord10 */
export type QualityControlRecord10Status = 'draft' | 'active' | 'pending' | 'completed' | 'cancelled' | 'archived';

export interface QualityControlRecord10 {
  id: string;
  externalCode: string;
  displayName: string;
  status: QualityControlRecord10Status;
  priority: number;
  facilityId: string;
  createdBy: string;
  updatedBy: string;
  notes: string;
  metadataJson: string;
  version: number;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
  archivedAt?: string | null;
}

export interface QualityControlRecord10CreateRequest {
  externalCode: string;
  displayName: string;
  status: QualityControlRecord10Status;
  priority: number;
  facilityId: string;
  notes?: string;
}

export interface QualityControlRecord10ListResponse {
  items: QualityControlRecord10[];
  total?: number;
}

export interface QualityControlRecord10StatsResponse {
  activeCount: number;
}

export function isQualityControlRecord10Active(rec: QualityControlRecord10): boolean {
  return rec.isActive && rec.status !== 'archived' && rec.status !== 'cancelled';
}

export function formatQualityControlRecord10Label(rec: QualityControlRecord10): string {
  return `${rec.externalCode} — ${rec.displayName}`;
}

export function compareQualityControlRecord10ByPriority(a: QualityControlRecord10, b: QualityControlRecord10): number {
  if (a.priority !== b.priority) return b.priority - a.priority;
  return a.displayName.localeCompare(b.displayName);
}

File: backend/internal/service/qualitycontrol_10.go
package service


import (
	"context"
	"time"

	"github.com/biruk-ak/clinical-lis/backend/internal/models"
	"github.com/biruk-ak/clinical-lis/backend/internal/repository"
	"github.com/google/uuid"
)

// QualityControlRecord10Service encapsulates business rules for QualityControlRecord10.
type QualityControlRecord10Service struct {
	repo *repository.QualityControlRecord10Repository
}

// NewQualityControlRecord10Service wires the service.
func NewQualityControlRecord10Service(repo *repository.QualityControlRecord10Repository) *QualityControlRecord10Service {
	return &QualityControlRecord10Service{repo: repo}
}

// Create validates and persists a new QualityControlRecord10.
func (s *QualityControlRecord10Service) Create(ctx context.Context, actor, facilityID, code, display, status string, priority int, notes string) (*models.QualityControlRecord10, error) {
	now := time.Now().UTC()
	rec := &models.QualityControlRecord10{
		ID:           uuid.NewString(),
		ExternalCode: code,
		DisplayName:  display,
		Status:       status,
		Priority:     priority,
		FacilityID:   facilityID,
		CreatedBy:    actor,
		UpdatedBy:    actor,
		Notes:        notes,
		MetadataJSON: "{}",
		Version:      1,
		IsActive:     true,
		CreatedAt:    now,
		UpdatedAt:    now,
	}
	if err := rec.Validate(); err != nil {
		return nil, err
	}
	if err := s.repo.Create(ctx, rec); err != nil {
		return nil, err
	}
	return rec, nil
}

// Get returns a QualityControlRecord10 by id.
func (s *QualityControlRecord10Service) Get(ctx context.Context, id string) (*models.QualityControlRecord10, error) {
	return s.repo.GetByID(ctx, id)
}

// List returns facility-scoped QualityControlRecord10 pages.
func (s *QualityControlRecord10Service) List(ctx context.Context, facilityID string, limit, offset int) ([]models.QualityControlRecord10, error) {
	if limit <= 0 || limit > 200 {
		limit = 50
	}
	if offset < 0 {
		offset = 0
	}
	return s.repo.List(ctx, facilityID, limit, offset)
}

// Update mutates an existing QualityControlRecord10.
func (s *QualityControlRecord10Service) Update(ctx context.Context, actor string, rec *models.QualityControlRecord10) error {
	if err := rec.Validate(); err != nil {
		return err
	}
	rec.UpdatedBy = actor
	rec.UpdatedAt = time.Now().UTC()
	// repository Update applies optimistic locking and increments version
	return s.repo.Update(ctx, rec)
}

// Archive soft-deletes a QualityControlRecord10.
func (s *QualityControlRecord10Service) Archive(ctx context.Context, id, actor string) error {
	return s.repo.Delete(ctx, id, actor)
}

// Search finds QualityControlRecord10 by external code prefix.
func (s *QualityControlRecord10Service) Search(ctx context.Context, facilityID, prefix string) ([]models.QualityControlRecord10, error) {
	return s.repo.SearchByCode(ctx, facilityID, prefix, 100)
}

// Stats returns active counts for dashboards.
func (s *QualityControlRecord10Service) Stats(ctx context.Context, facilityID string) (int64, error) {
	return s.repo.CountActive(ctx, facilityID)
}
